optuna/storages/redis_storage.py

```diff
@@ -0,0 +1,511 @@
+import copy
+from datetime import datetime
+import pickle
+import redis
+import threading
+
+from optuna import distributions, exceptions  # NOQA
+from optuna.storages import base
+from optuna.storages.base import DEFAULT_STUDY_NAME_PREFIX
+from optuna import structs
+from optuna import type_checking
+
+if type_checking.TYPE_CHECKING:
+    from typing import Any  # NOQA
+    from typing import Dict  # NOQA
+    from typing import List  # NOQA
+    from typing import Optional  # NOQA
+
+
+class RedisStorage(base.BaseStorage):
```

Comment:
To allow multiple processes accessing the same Redis database (similar to our other database. See https://optuna.readthedocs.io/en/latest/tutorial/distributed.html) we might have to rely on Redis' `WATCH` (e.g. https://redis-py.readthedocs.io/en/latest/index.html#redis.Redis.pipeline, https://redis-py.readthedocs.io/en/latest/index.html#redis.Redis.transaction). We could tackle that in a separate PR since this change is already quite big. In that case, we might want to mark this feature as experimental.

Reply:
Added pipeline blocks in places where multiple updates where performed. That will pipeline all writes as one atomic transaction.